hail/src/main/scala/is/hail/expr/types/physical/PCanonicalStruct.scala

```diff
@@ -0,0 +1,268 @@
+
+          (updateKey(key, j, insertedFieldType), { (region, offset, rvb, inserter) =>
+            rvb.startStruct()
+            var i = 0
+            while (i < j) {
+              if (region != null)
+                rvb.addField(this, region, offset, i)
+              else
+                rvb.setMissing()
+              i += 1
+            }
+            if (region != null && isFieldDefined(region, offset, j))
+              fieldInserter(region, loadField(region, offset, j), rvb, inserter)
+            else
+              fieldInserter(null, 0, rvb, inserter)
+            i += 1
+            while (i < localSize) {
+              if (region != null)
+                rvb.addField(this, region, offset, i)
+              else
+                rvb.setMissing()
+              i += 1
+            }
+            rvb.endStruct()
+          })
+
+        case None =>
+          val (insertedFieldType, fieldInserter) = PCanonicalStruct(Array.empty[PField], required).unsafeInsert(typeToInsert, path.tail)
+
+          (appendKey(key, insertedFieldType), { (region, offset, rvb, inserter) =>
+            rvb.startStruct()
+            var i = 0
+            while (i < localSize) {
+              if (region != null)
+                rvb.addField(this, region, offset, i)
+              else
+                rvb.setMissing()
+              i += 1
+            }
+            fieldInserter(null, 0, rvb, inserter)
+            rvb.endStruct()
+          })
+      }
+    }
+  }
+
+  def updateKey(key: String, i: Int, sig: PType): PStruct = {
+    assert(fieldIdx.contains(key))
+
+    val newFields = Array.fill[PField](fields.length)(null)
+    for (i <- fields.indices)
+      newFields(i) = fields(i)
+    newFields(i) = PField(key, sig, i)
+    PCanonicalStruct(newFields, required)
+  }
+
+  def deleteField(key: String): PStruct = {
+    assert(fieldIdx.contains(key))
+    val index = fieldIdx(key)
+    if (fields.length == 1)
+      PCanonicalStruct(Array.empty[PField], required)
+    else {
+      val newFields = Array.fill[PField](fields.length - 1)(null)
+      for (i <- 0 until index)
+        newFields(i) = fields(i)
+      for (i <- index + 1 until fields.length)
+        newFields(i - 1) = fields(i).copy(index = i - 1)
+      PCanonicalStruct(newFields, required)
+    }
+  }
+
+  def appendKey(key: String, sig: PType): PStruct = {
+    assert(!fieldIdx.contains(key))
+    val newFields = Array.fill[PField](fields.length + 1)(null)
+    for (i <- fields.indices)
+      newFields(i) = fields(i)
+    newFields(fields.length) = PField(key, sig, fields.length)
+    PCanonicalStruct(newFields, required)
+  }
+
+
+  def rename(m: Map[String, String]): PStruct = {
+    val newFieldsBuilder = new ArrayBuilder[PField]()
+    var idx = 0
+    fields.foreach { fd =>
+      val n = fd.name
+      newFieldsBuilder += PField(m.getOrElse(n, n), fd.typ, idx)
+      idx += 1
+    }
+    PCanonicalStruct(newFieldsBuilder.result(), required)
```

Comment:
note that required is now being passed, [previously was not, seems like a bug](https://github.com/hail-is/hail/pull/7733/files#diff-bf439d99cf3638d4e2e1c9f6be564171L178)